[Microsoft.Workoads] Initialize connectors for 2023-10-01-preview api version

nishanthm-msft opened the pull request:
## Changes (Please Note)
- **PR with changes [already reviewed and merged in Private repo](https://github.com/Azure/azure-rest-api-specs-pr/pull/14463). This swagger PR is copy of already merged API version in private repo and this PR replicates the same change from private repo to public repo.**

- There are no new swagger model changes in this PR when compared to the private repo api version swagger model.

- Net new changes in this PR to resolve Swagger LintDiff checks differing from private repo version (There were no Swagger LintDiff errors in private repo but came up in this public repo) are:
   - Removed 200 response for Delete LROs as indicated by the Swagger LintDiff check
   - Updated response status code from 201 to 202 for Patch LROs as indicated by the Swagger LintDiff check
   - Added Regex naming patterns for parameters as indicated by the Swagger LintDiff check

- Introduced [alternate folder structure](https://github.com/Azure/azure-rest-api-specs#folder-structure-for-service-group) for Microsoft.Workloads/connectors RT.
     - microsoft.workloads/connectors -> Tracked resource
     - microsoft.workloads/connectors/acssBackups -> Tracked resource

### Notes

- This new RT is following [alternate folder structure](https://github.com/Azure/azure-rest-api-specs#folder-structure-for-service-group), so this was not possible to be created using [OpenAPI HUB](https://openapihub.azure-devex-tools.com/) when creating new RT in existing RP. hence commits order might not be as expected from OpenAPI HUB. 
- This is initial release for this RT with alternate folder structure. Same API version was merged in private repo (https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview)

# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/b321e1dc-5e07-4940-9cd6-a65ac6729a8b)

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

Commit:
Initialize connectors for 2023-10-01-preview api verison

Commit:
add unknown words to custom-words

Commit:
fix linter checks

Commit:
additional fixes for linter checks

Commit:
remove 200 response for delete LROs

Commit:
update examples to remove 200 from delete LROs

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/common-types/v1/commonTypes.json:
```diff
@@ -0,0 +1,69 @@
+      ]
+    }
+  },
+  "parameters": {}
```
Why are these being redefined? Shouldn't they be referred directly from the common types? Should this file exist at all?

nishanthm-msft replied:
This commonTypes file acts as a wrapper on top of actual [commonTypes ](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/common-types) since we have custom properties to be included for User MSI input. This file actually references the original commonTypes as done in Line 60

pareshverma91 replied:
The type you are trying to create here seems to be there already as ManagedServiceIdentity [https://github.com/Azure/azure-rest-api-specs/blob/077f1f8f0c8f7e5ac2ab17f9b80d36c56dc5156c/specification/common-types/resource-management/v5/managedidentity.json#L50](https://github.com/Azure/azure-rest-api-specs/blob/077f1f8f0c8f7e5ac2ab17f9b80d36c56dc5156c/specification/common-types/resource-management/v5/managedidentity.json#L50). It seems you are removing fields instead of adding.

nishanthm-msft replied:
For our service in ManagedServiceIdentityType, only None, UserAssigned types are allowed whereas in actual commonTypes both SystemAssigned and UserAssigned types are allowed. For this purpose, we have defined our own ManagedServiceIdentityType definition. Since only UserAssigned type is allowed in our service we have created definition for that as well and inside it we refer to actual commonTypes file for the common properties to reuse instead of redefining those in our swagger

nishanthm-msft commented:
- **PR with changes [already reviewed and merged in Private repo](https://github.com/Azure/azure-rest-api-specs-pr/pull/14463). This swagger PR is copy of already merged API version in private repo and this PR replicates the same change from private repo to public repo.**

- **There are no new swagger model changes in this PR when compared to the private repo api version swagger model**.

- Net new changes in this PR to resolve Swagger LintDiff checks differing from private repo version (There were no Swagger LintDiff errors in private repo but came up in this public repo) are:
   - Removed 200 response for Delete LROs as indicated by the Swagger LintDiff check
   - Updated response status code from 201 to 202 for Patch LROs as indicated by the Swagger LintDiff check
   - Added Regex naming patterns for parameters as indicated by the Swagger LintDiff check

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/backupTypes.json:
```diff
@@ -0,0 +1,849 @@
+      "properties": {
+        "schedulePolicyType": {
+          "description": "This property will be used as the discriminator for deciding the specific types in the polymorphic chain of types.",
+          "type": "string"
```
>"type": "string" [](http://example.com/codeflow?start=10&length=16)

Should this be an enum?

nishanthm-msft replied:
We need this property to be a string. Azure Backup service has defined this property as string https://github.com/Azure/azure-rest-api-specs/blob/077f1f8f0c8f7e5ac2ab17f9b80d36c56dc5156c/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json#L9152. Our service is integrated with Azure Backup service. Currently we have defined the backup policy models in our swagger. But before going GA, we would reference the policy models from Azure Backup swagger and not have ownership from our side.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/backupTypes.json:
```diff
@@ -0,0 +1,849 @@
+        "retentionPolicyType": {
+          "description": "This property will be used as the discriminator for deciding the specific types in the polymorphic chain of types.",
+          "type": "string"
+        }
```
Should this be an enum?

nishanthm-msft replied:
We need this property to be a string. Azure Backup service has defined this property as string https://github.com/Azure/azure-rest-api-specs/blob/077f1f8f0c8f7e5ac2ab17f9b80d36c56dc5156c/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json#L9152. Our service is integrated with Azure Backup service. Currently we have defined the backup policy models in our swagger. But before going GA, we would reference the policy models from Azure Backup swagger and not have ownership from our side.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/backupTypes.json:
```diff
@@ -0,0 +1,849 @@
+        "makePolicyConsistent": {
+          "description": "Fix the policy inconsistency",
+          "type": "boolean"
+        }
```
Typical recommendation is to replace boolean with enum offering extensibility. However, not stressing on this since this has been reviewed in the previous private repo merge.

nishanthm-msft commented:
Hi @pareshverma91 can you please provide a sign off if the changes are good to be merged?

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/backupTypes.json:
```diff
@@ -0,0 +1,849 @@
+          "description": "SQL compression flag",
+          "type": "boolean"
+        },
+        "isCompression": {
```
>isCompression [](http://example.com/codeflow?start=9&length=13)

You cannot contribute to repositories outside of your enterprise Microsoft EMU

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/backupTypes.json:
```diff
@@ -0,0 +1,849 @@
+          "description": "SQL compression flag",
+          "type": "boolean"
+        },
+        "isCompression": {
```
>isCompression [](http://example.com/codeflow?start=9&length=13)

A setting named isCompression sounds off. Did you mean this to be an enum Compression with values Enabled/Disabled or a boolean compressionEnabled?

nishanthm-msft replied:
Our service is integrated with Azure Backup service. Currently we have defined the backup policy models in our swagger. But before going GA, we would reference the policy models from Azure Backup swagger and not have ownership from our side.
So to avoid renaming of properties in future when we reference these definitions directly from Azure backup service swagger, we have replicated the definitions as how they are in their swagger.

https://github.com/Azure/azure-rest-api-specs/blob/5797d78f04cd8ca773be82d2c99a3294009b3f0a/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json#L9169

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/backupTypes.json:
```diff
@@ -0,0 +1,849 @@
+          "$ref": "#/definitions/UserAssignedIdentityProperties"
+        }
+      }
+    },
```
In other comment you called out the api restriction because of which you have had to do it like so. Please note the impact of this would be that customers when using the SDK might have to manually construct the object from the typical resource used across rest of the Azure.
Considering it as non-blocking for now.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+        "description": "Creates a connector resource",
+        "parameters": [
+          {
+            "$ref": "../../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
```
>v3 [](http://example.com/codeflow?start=72&length=2)

In commonTypes.json, you are referring to v5. Standardizing on same version of common types would be better.

nishanthm-msft replied:
Microsoft.Workloads/commonTypes.json is common across the Workloads platform where other workloads swaggers also use this apart from our Connector service. This is the reason that there is a difference in actual commonTypes version being referenced.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+                "description": "URL to query for status of the operation.",
+                "type": "string"
+              }
+            }
```
Instead of this, shouldn't [https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation) and [https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options) suffice?

nishanthm-msft replied:
In private repo PR for the same swagger, there was a comment (https://github.com/Azure/azure-rest-api-specs-pr/pull/14463#discussion_r1313254404) suggesting that https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options is not required for Control plane operations and only https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation is enough. Have kept that only based on that suggestion

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+        "Updating",
+        "Creating",
+        "Failed",
+        "Deleting",
```
>Deleting [](http://example.com/codeflow?start=9&length=8)

Should there be a corresponding Deleted state? Or is the expectation the resource won't exist once Deleted and hence the state isn't needed.

nishanthm-msft replied:
Yes, the connector parent resource and connection child resource won't exist once Deleted and hence Deleted state wasn't required

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+      }
+    },
+    "NewRecoveryServicesVault": {
+      "x-ms-discriminator-value": "New",
```
>New [](http://example.com/codeflow?start=35&length=3)

Please do review some of the generated SDKs. I believe the discriminators are resolved globally and not within the context of local types. This would imply a resource with discriminator value of New would conflict with any other resource definition that uses the discriminator value of New, even if it is a different type.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+        }
+      }
+    },
+    "RecoveryServicesVaultProperties": {
```
>RecoveryServicesVaultProperties [](http://example.com/codeflow?start=5&length=31)

Its child types are *ServicesVault and its usage/reference elsewhere makes it sound like that there shouldn't be a Properties suffix. Please review one of the generated SDKs to see whether the generated class names look reasonable.
Marking the comment as non-blocking though.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+          }
+        },
+        "isInclusionList": {
+          "description": "Flag to indicate whether DiskLunList is to be included/ excluded from backup.",
```
Would the name includeInBackup be more apt?

nishanthm-msft replied:
Similar to the other comment this definition is replicated from Azure backup service swagger https://github.com/Azure/azure-rest-api-specs/blob/5797d78f04cd8ca773be82d2c99a3294009b3f0a/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json#L6532.
Keeping the same definitions as of their swagger would help prevent doing service changes to handle property renames when we reference their swagger directly

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+            "openssl"
+          ],
+          "x-ms-enum": {
+            "name": "sslCryptoProvider",
```
>sslCryptoProvider [](http://example.com/codeflow?start=21&length=17)

I wonder if this should be SslCryptoProvider.

nishanthm-msft replied:
Since we have followed this naming convention of properties names starting with lower case in all the definitions, keeping it same here as well

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+        }
+      }
+    },
+    "ConnectorErrorDefinition": {
```
>ConnectorErrorDefinition [](http://example.com/codeflow?start=5&length=24)

Could you please review the ErrorDetail type in common types? I wonder if that is suitable here.

nishanthm-msft replied:
This was custom defined during the inception of the service to limit the properties required to be populated and to enhance it as required in future.

pareshverma91 commented on specification/workloads/resource-manager/Microsoft.Workloads/connectors/preview/2023-10-01-preview/connectors.json:
```diff
@@ -0,0 +1,1098 @@
+        "read"
+      ],
+      "x-ms-enum": {
+        "name": "backupType",
```
>backupType [](http://example.com/codeflow?start=17&length=10)

I wonder if this should be BackupType.

nishanthm-msft replied:
Since we have followed this naming convention of properties names starting with lower case in all the definitions, keeping it same here as well

pareshverma91 commented:
Left a few additional comments. However, signing off based on no egregious issues and to unblock the final review.

nishanthm-msft commented:
> Left a few additional comments. However, signing off based on no egregious issues and to unblock the final review.

Thanks @pareshverma91. Addressed the comments/suggestions provided. Thanks for the sign-off

nishanthm-msft commented:
/pr RequestMerge
